### PR TITLE
Use HDFS when storing resources for tethered runs

### PR DESCRIPTION
**Pull request by nliu132**
- Update TetheringAgent to write to file system and store the URI in program options
- Update `DistributedProgramRunner` to read from file system for tethered runs and localize the files

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/ProgramOptionConstants.java [hunk 1]
```diff
@@ -159,4 +159,9 @@
    * artifact fetching
    */
   public static final String PEER_NAMESPACE = "peerNamespace";
+
+  /**
+   * Option for URI to additional resources needed for program run. This is needed for tethered programs
+   */
+  public static final String PROGRAM_RESOURCE_URI = "programResourceUri";
```
Comment (rmstar) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/ProgramOptionConstants.java:167]: nit: "Option for URI to additional resources needed for program run" -> missing a word?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/DistributedProgramRunner.java [hunk 2]
```diff
@@ -387,27 +391,18 @@ public final ProgramController run(final Program program, ProgramOptions oldOpti
...
+  private void loadAdditionalResources(Arguments args, CConfiguration cConf, File tempDir) throws IOException {
+    Location location = locationFactory.create(args.getOption(ProgramOptionConstants.PROGRAM_RESOURCE_URI));
+    // add additional cConf entries
+    File tetherCConfFile = File.createTempFile("cdap-tether", ".xml", tempDir);
+    try (InputStream is = location.append(TETHER_CONF_FILE_NAME).getInputStream()) {
+      Files.copy(is, tetherCConfFile.toPath(), StandardCopyOption.REPLACE_EXISTING);
+      LOG.debug("Copied {} to {}", TETHER_CONF_FILE_NAME, tetherCConfFile);
     }
+    cConf.addResource(tetherCConfFile.toURI().toURL());
+    cConf.reloadConfiguration();
```
Comment (rmstar) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/DistributedProgramRunner.java]: Do we need to call cConf.reloadConfiguration()? Looks like it's called as part of addResource() [here](https://github.com/cdapio/cdap/blob/433f41d8e682ba5d4592b6ec0d139c64970e3c40/cdap-common/src/main/java/io/cdap/cdap/common/conf/Configuration.java#L460)

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/DistributedProgramRunner.java [hunk 3]
```diff
@@ -604,7 +599,18 @@ private ProgramOptions updateProgramOptions(ProgramOptions options,
    * classpath.
    */
   @Nullable
-  private URI getLogBackURI(Program program) throws URISyntaxException {
+  private URI getLogBackURI(Program program, Arguments args, File tempDir) throws URISyntaxException, IOException {
```
Comment (rmstar) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/DistributedProgramRunner.java:601]: Not directly related to these changes - do you if it's sufficient to handle logback.xml, or do we also care about logback-container.xml (see L624 below)?

**Review by rmstar — APPROVED** [review]

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/DistributedProgramRunner.java (hunk 2 above)
Comment (nliu132) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/DistributedProgramRunner.java]: looks like the call is unneeded. removed

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/ProgramOptionConstants.java (hunk 1 above)
Comment (nliu132) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/ProgramOptionConstants.java:167]: reworded to sound less awkward

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/DistributedProgramRunner.java (hunk 3 above)
Comment (nliu132) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/DistributedProgramRunner.java:601]: Based on [the docs](https://cdap.atlassian.net/wiki/spaces/DOCS/pages/480346897/Application+Logback), it sounds like `logback-container.xml` is the default configuration and `logback.xml` should be used when it exists

**Pull request merged**